xpc_basic/man/stringmap.1
.\"******************************************************************************
.\" xpc_suite/xpc/man/stringmap.1
.\"------------------------------------------------------------------------------
.\"
.\" \file       	stringmap.1
.\" \library    	xpc++
.\" \author     	Chris Ahlstrom
.\" \date       	2010-05-30 to 2010-05-30
.\" \version    	$Revision$
.\" \license    	$XPC_SUITE_GPL_LICENSE$
.\"
.\" Hey, EMACS: -*- nroff -*-
.\"
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.\"
.\" Please adjust the following date whenever revising the manpage.
.\"
.\"------------------------------------------------------------------------------

.TH XPCCUT 1.0.4 "May 30, 2010"

.\"******************************************************************************
.\" Some roff macros, for reference:
.\"
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\"
.\" for manpage-specific macros, see man(7)
.\"
.\"------------------------------------------------------------------------------

.de BP
.sp
.ti \-.2i
\(**
..

.SH NAME
STRINGMAP \(em\& stringmap -- An easy way to handle sets of strings.

.SH SYNOPSIS
xxx

.SH DESCRIPTION

This man page is not well-maintained and is incomplete.
The Doxygen documentation in
/usr/share/doc/xpc++-1.0/html/index.html
or
/usr/local/share/doc/xpc++-1.0/html/index.html
is the authoritative source of XPC++ information.

.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invoke bold face and italics,
.\" respectively.

\fBxpc++\fP is a library that provides a simple
framework for C++ programs.  The \fBstringmap\fP module
provides a way to handle sets of strings

.SH OPTIONS
.I xpc++
options are discussed thoroughly in the Doxygen documentation.
The unit-test application can be run with the --help option to see them all.

.TP
.B \-h --help
Print an informative help message describing the options and then exit.

.TP
.B \-v --version
Print the version number of
.I unit_test_test
on the standard error output and then exit.

.SH "BUGS"
There are bugs, of course.  But so far our unit-testing has found them all.
Ha ha.

.SH "AUTHORS"
.I xpc++
was written by Chris Ahlstrom.
This man page was written by Chris Ahlstrom.

.SH "SEE ALSO"
.I The C Programming Language
by B. W. Kernighan and D. M. Ritchie, Prentice-Hall, New Jersey, 1978.

.\"******************************************************************************
.\" xpc_suite/xpc++/man/stringmap.1
.\"------------------------------------------------------------------------------
.\" Local Variables:
.\" End:
.\"------------------------------------------------------------------------------
.\" vim: ts=3 sw=3 ft=nroff
.\"------------------------------------------------------------------------------


xpc_basic/xpc++/src/stringmap.cpp
/******************************************************************************
 * stringmap.cpp
 *------------------------------------------------------------------------*//**
 *
 * \file          stringmap.cpp
 * \library       xpc
 * \author        Chris Ahlstrom
 * \date          2010-05-30
 * \updates       2014-04-21
 * \version       $Revision$
 * \license       $XPC_SUITE_GPL_LICENSE$
 *
 *    This module defines an std::map that uses only std::string as a key
 *    value, for easy lookup.
 *
 *//*-------------------------------------------------------------------------*/

#include <xpc/errorlogging.h>          /* C::xpc_logfile()                    */
#include <xpc/stringmap.hpp>           /* xpc::stringmap<VALUETYPE>           */
XPC_REVISION(stringmap)                /* show_stringmap_info()               */

namespace xpc
{

/******************************************************************************
 * iequal()
 *------------------------------------------------------------------------*//**
 *
 *    A simple case-insensitive comparison for simple string equality.
 *
 *    It is a bit similar to the cmp_nocase() function in section 20.3.8
 *    "Comparisons" of Stroustrop's C++ Programming Language book.
 *
 *    We just stick it in this existing module for convenience.
 *
 * \note
 *    If your application is already using Boost, you can use the
 *    boost::iequal() function instead.
 *
 * \param s1
 *    Provides the first string.
 *
 * \param s2
 *    Provides the second string.
 *
 * \return
 *    Returns true of the strings are identical except for case.
 *
 *//*-------------------------------------------------------------------------*/

bool
iequal (const std::string & s1, const std::string & s2)
{
   bool result = s1.size() == s2.size();
   if (result)
   {
      std::string::const_iterator p1 = s1.begin();
      std::string::const_iterator p2 = s2.begin();
      while (p1 != s1.end() && p2 != s2.end())
      {
         if (toupper(*p1) != toupper(*p2))
         {
            result = false;
            break;
         }
         else
         {
            p1++;
            p2++;
         }
      }
   }
   return result;
}

/******************************************************************************
 * show()
 *------------------------------------------------------------------------*//**
 *
 *    Writes out the contents of the string to standard output, in a
 *    stylized format.
 *
 *    Also see the show_pair() and show() functions in the stringmap.hpp
 *    module.
 *
 * \param tag
 *    A name for the string, to provide context for the human reader.
 *
 * \param s
 *    The string to be shown.
 *
 *//*-------------------------------------------------------------------------*/

void
show (const std::string & tag, const std::string & s)
{
   fprintf
   (
      stdout,
      "-    String ('%8s')       '%s'\n",
      tag.c_str(), s.c_str()
   );
}

}        // namespace xpc

/******************************************************************************
 * stringmap.cpp
 *-----------------------------------------------------------------------------
 * Local Variables:
 * End:
 *-----------------------------------------------------------------------------
 * vim: ts=3 sw=3 et ft=cpp
 *----------------------------------------------------------------------------*/
